### PR TITLE
Additional JSON manipulation functions

### PR DESCRIPTION
**Pull request by PhictionalOne**
The current state of JSON manipulation in the JSON extension is rather limited.
In order to modify existing JSON arrays and objects this PR aims to introduce the following functions that work directly on JSON:

| function name         | Description                                       |
|:----------------------|:--------------------------------------------------|
| `json_array_insert`   | Inserts an element at a given index in JSON Array |
| `json_array_append`   | Adds an element at the end                        |
| `json_array_prepend`  | Adds an element to the the front                  |
| `json_array_remove`   | Removes element(s) at index/in index range        |
| `json_array_rotate`   | Rotates JSON Array by amount $n$                  |
| `json_array_take`     | Take the first $n$ elements from JSON Array       |
| `json_array_tail`     | Drops the first element of JSON Array             |
| `json_obj_add`        | Adds a value by a given key to JSON object        |
| `json_obj_rename_key` | Renames a key in an JSON object                   |
| `json_obj_replace`    | Replaces a value by a given key                   |

The simple behaviour of the JSON array specific functions is the following:

```sql
D SELECT json_array_insert(arr, '"a"', 2) AS "insert",
         json_array_append(arr, '"a"')    AS "append",
         json_array_prepend('"a"', arr)   AS "prepend",
         json_array_remove(arr, 2)        AS "remove",
         json_array_rotate(arr, 1)        AS "rotate",
         json_array_take(arr, 1)          AS "take",
         json_array_tail(arr)             AS "tail"
  FROM   (VALUES ('[54,64,10]')) AS _(arr);
+----------------+----------------+----------------+---------+------------+------+---------+
|     insert     |     append     |    prepend     | remove  |   rotate   | take |  tail   |
+----------------+----------------+----------------+---------+------------+------+---------+
| [54,64,"a",10] | [54,64,10,"a"] | ["a",54,64,10] | [54,64] | [64,10,54] | [54] | [64,10] |
+----------------+----------------+----------------+---------+------------+------+---------+
```

It should be noted that negative indeces index from the back or switch the rotation direction.

Whereas the behaviour for the three JSON object functions is like so:

```sql
D SELECT json_obj_add(obj, 'new', 42::INT)    AS "add",
         json_obj_rename_key(obj, 'b', 'new') AS "rename key",
         json_obj_replace(obj, 'a', 42::INT)  AS "replace"
  FROM   (VALUES ('{"a": true, "b": 1337}')) AS _(obj);
+------------------------------+-----------------------+-------------------+
|             add              |      rename key       |      replace      |
+------------------------------+-----------------------+-------------------+
| {"a":true,"b":1337,"new":42} | {"a":true,"new":1337} | {"a":42,"b":1337} |
+------------------------------+-----------------------+-------------------+
```

An extended documentation can be found in the corresponding [PR](https://github.com/duckdb/duckdb-web/pull/2992).

### CODE REVIEW

File: extension/json/include/json_executors.hpp [hunk 1]
```diff
@@ -29,6 +29,31 @@ struct JSONExecutors {
...
+			auto doc = JSONCommon::ReadDocument(input, JSONCommon::READ_FLAG, alc);
+
+			// immutable document to mutable document
+			auto mut_doc = yyjson_doc_mut_copy(doc, alc);
+
+			// compute mutable result
+			auto new_val = fun(mut_doc->root, alc, result);
+
+			// Convert mutable value back to immutable document
+			doc = yyjson_mut_val_imut_copy(new_val, alc);
```
Comment (Maxxen) [inline extension/json/include/json_executors.hpp:50]: Can't you avoid this copy by just calling `WriteVal<yyjson_mut_val>(doc->root, alc)` instead?

File: extension/json/json_functions/json_array_append.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+#include "json_executors.hpp"
+
+namespace duckdb {
+
+//! Append String or JSON value to an array
+yyjson_mut_val *ArrayAppendJSON(yyjson_mut_val *arr, string_t element, yyjson_alc *alc, Vector &result) {
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_append.cpp:6]: This (and all the other operation "implementations") can be static (or maybe even a lambda, although I wouldn't change that now unless you really want to)

File: extension/json/json_functions/json_array_append.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+static void ArrayAppendFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	auto left_type = args.data[0].GetType();
+	D_ASSERT(left_type == LogicalType::VARCHAR || left_type == LogicalType::JSON());
+	auto right_type = args.data[1].GetType();
+	D_ASSERT(right_type == LogicalType::VARCHAR || right_type == LogicalType::JSON());
+
+	JSONExecutors::BinaryMutExecute<string_t>(args, state, result, ArrayAppendJSON);
+}
+
+static void GetArrayAppendFunctionInternal(ScalarFunctionSet &set, const LogicalType &lhs, const LogicalType &rhs) {
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_append.cpp:28]: Not really important so feel free to ignore, but does this really necessitate a new function?

File: extension/json/json_functions/json_array_rotate.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	size_t array_size = yyjson_mut_arr_size(arr);
+	if (amount == 0 || array_size == 0) {
+		return arr;
+	}
+
+	size_t shift_amount = 0;
+
+	if (amount < 0) {
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_rotate.cpp:19]: You want to bounds-check this and use `(Unsafe)NumericCast<AMOUNT_TYPE, size_t>` instead. You could maybe make use of `TryAddOperator` and `TrySubtractOperator` potentially, or just catch the internal exception thrown when NumericCast fails. `sizeof(int64_t) != sizeof(size_t)` on e.g. WASM

File: extension/json/json_common.cpp [hunk 5]
```diff
@@ -334,4 +334,21 @@ void JSONCommon::GetWildcardPath(yyjson_val *val, const char *ptr, const idx_t &
...
+	if (idx < 0) {
+		idx += array_size;
+	}
+
+	// clamp to minimum 0
+	if (idx < 0) {
+		idx = 0;
+	}
+
+	return static_cast<size_t>(idx);
```
Comment (Maxxen) [inline extension/json/json_common.cpp:351]: Don't static cast to size_t, see my other comment.

File: extension/json/json_functions/json_array_remove.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	if (!yyjson_mut_is_arr(arr)) {
+		throw InvalidInputException("JSON input not an JSON Array");
+	}
+
+	if (start < 0 || end < 0 || start > end) {
+		throw InvalidInputException("Invalid range indices");
+	}
+
+	size_t array_length = yyjson_mut_arr_size(arr);
+	size_t index = static_cast<size_t>(start);
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_remove.cpp:36]: don't static_cast

File: extension/json/json_functions/json_array_take.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		throw InvalidInputException("JSON input not an JSON Array");
+	}
+
+	if (n <= 0) {
+		// TODO: Maybe introduce negative space to take from the end?
+		auto doc = JSONCommon::CreateDocument(alc);
+		return yyjson_mut_arr(doc);
+	}
+
+	size_t last_index = static_cast<size_t>(n);
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_take.cpp:17]: static_cast

File: extension/json/json_functions/json_array_take.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	if (!yyjson_mut_is_arr(arr)) {
+		throw InvalidInputException("JSON input not an JSON Array");
+	}
+
+	if (start < 0 || end < 0 || start > end) {
+		throw InvalidInputException("Invalid range indices");
+	}
+
+	size_t array_length = yyjson_mut_arr_size(arr);
+	size_t first_index = static_cast<size_t>(start);
```
Comment (Maxxen) [inline extension/json/json_functions/json_array_take.cpp:39]: static_cast

**Review by Maxxen — CHANGES_REQUESTED** [review]
Looks great! Some comments below:

**Comment by Mytherin** [comment]
Thanks for the contribution!

Generally when implementing new functionality we try to look if the same functionality already exists in other systems - in particular in Postgres. For example - both [SQLite](https://sqlite.org/json1.html#jins) and [Postgres](https://www.postgresql.org/docs/current/functions-json.html) seem to support functions like `json(b)_insert`. Looking at the functions here - this seems like a more flexible variant of these functions (but also much more complex to implement of course). For example:

```sql
-- json_array_insert
select jsonb_insert('[1,2,3,4]', '{2}', '99');
--  [1, 2, 99, 3, 4]

-- json_array_append
select jsonb_insert('[1,2,3,4]', '{-1}', '99', true);
--  [1, 2, 3, 4, 99]

-- json_array_prepend
select jsonb_insert('[1,2,3,4]', '{0}', '99');
-- [99, 1, 2, 3, 4]

-- json_array_remove
select '[1,2,3,4]' #- '{2}';
--  [1, 2, 4]
```

I'm a bit hesitant to add a number of functions that do very specific things, when we could be adding a function that supports all of these operations. For example what happens if users want to append to a nested array or to an object? `jsonb_insert` supports this through the following syntax, but this would not be possible in the current more specialized function: 

```sql
select jsonb_insert('[1,[2,3],4]', '{1, 2}', '99');
-- [1, [2, 3, 99], 4]

```

@lnkuiper perhaps you have any more thoughts here?

File: extension/json/include/json_executors.hpp (hunk 1 above)
Comment (PhictionalOne) [reply extension/json/include/json_executors.hpp]: thanks, done

**Comment by PhictionalOne** [comment]
> Generally when implementing new functionality we try to look if the same functionality already exists in other systems - in particular in Postgres. For example - both [SQLite](https://sqlite.org/json1.html#jins) and [Postgres](https://www.postgresql.org/docs/current/functions-json.html) seem to support functions like `json(b)_insert`. [This] seems like a more flexible variant of these functions (but also much more complex to implement of course).

Hmm.. Yes you're right the `jsonb_insert` is indeed more versatile. SQLite and Postgres use their own notation for "pathing". Considering that `yyjson` has the functionality for [JSON Pointer](https://datatracker.ietf.org/doc/html/rfc6901) the `json_insert` looked like this after I implementation:

```sql
D SELECT json_insert('[1,2,[5,6],4]', '/2/1', '"hello"');
┌───────────────────────────────────────────────────────┐
│ json_insert('[1,2,[5,6],4]', '/2/1', '"hello"')       │
│                         json                          │
├───────────────────────────────────────────────────────┤
│ [1,2,[5,"hello",6],4]                                 │
└───────────────────────────────────────────────────────┘
```

A downside of JSON Pointer in that scenario is that there are no negative indeces. (You could do some string operations and stepwise descend by calculating the _positive_ index though)

This `json_insert` would replace my proposed `json_array_prepend`, `json_obj_add` and `json_array_insert` with the caveat of no indexing from the end.

Using this approach I could implement a `json_remove` in a similar fashion and also retrofit most of the other function with a pointer argument as well.

Is that something that would be more in line with the SQLite and Postgres behaviour?

**Comment by lnkuiper** [comment]
@PhictionalOne It makes sense to use the JSON Pointer to insert, this looks better.

We also support the `$.field[array_index]` syntax which allows indexing from the back. Both are implemented for immutable yyjson, but could be templated to work for mutable yyjson as well. You could consider adding that to this PR or leaving it for a future PR.

**Comment by kylejbrk** [comment]
Is this feature still going to be added? i cant find any current way to update an existing json field using a path.

**Comment by lnkuiper** [comment]
Hi @kylejbrk, there are many JSON functions that other systems support that we would also like to add to DuckDB, but they are not high on the (long) priority list

**Pull request closed without merging**